### PR TITLE
Feature/Update Poll Status Based on Minimum Approval and Quorum Criteria

### PR DESCRIPTION
**Pull request by jayapal**
#### Summary

- **Docker Setup**: Added `Dockerfile` and `docker-compose.yml` to simplify the code setup, allowing services to be spun up with a single command.
- **Makefile**: Introduced a `Makefile` to further streamline managing the Docker environment.

#### Poll Model Enhancements

- **New Fields**:
  1. `approval_minimum`: Required, default value 51, range 0-100.
  2. `finalization_period`: Required, default value 3, range 1-30.
  3. `finalization_period_start_date`: Optional, records the date when the poll status changes to **Finalization Period**.
  4. **Existing Field Update**: `quorum` is now mandatory with a default value of 51.

- **Poll Status**: Added a new status option '2' for **Finalization Period**.

#### Serializer and Function Updates

- **PollSerializer**: Updated to reflect changes in the Poll model.
- **Approval and Quorum Check**:
  - Added `check_poll_met_approval_and_quorum` function in `services.py` within the proposal module.
  - This function checks if a poll has met the minimum approval and quorum criteria and is invoked whenever a user updates the priority of a poll proposal.
  - If the poll meets the criteria, its status is changed to **Finalization Period**. If a user downvotes and the criteria are no longer met, the status reverts to **Ongoing** (0).

#### Periodic Task

- **Finalization Check Task**: Introduced a new periodic Celery task that scans all polls in the **Finalization Period**. It compares the current date with `finalization_period_start_date` to determine if the poll has reached its finalization period. If so, the poll status is set to **Finished** (1).
- The task is scheduled using the Django admin panel.

#### Testing

- **Test Cases**: Added test cases for all the new additions.

### CODE REVIEW

File: flowback/poll/services/poll.py [hunk 1]
```diff
@@ -301,3 +308,50 @@ def poll_priority_update(user_id: int, poll_id: int, score: int) -> None:
...
+
+    try:
+        proposal = PollProposal.objects.get(id=proposal_id)
+        poll = Poll.objects.get(id=proposal.poll_id)
+    except (PollProposal.DoesNotExist, Poll.DoesNotExist):
+        print('Poll proposal or poll does not exist')
+        return False
+    
+    poll_community = poll.created_by.group
+    total_community_members = Group.objects.get(id=poll_community.id).groupuser_set.count()
```
Comment (Gofven) [inline flowback/poll/services/poll.py]: Omit people with active=False

File: flowback/poll/services/poll.py [hunk 2]
```diff
@@ -301,3 +308,50 @@ def poll_priority_update(user_id: int, poll_id: int, score: int) -> None:
...
+    total_voted_community_members_percentage = (proposal.votes.count() / total_community_members) * 100
+    print('Total voted community members percentage:', total_voted_community_members_percentage)
+
+    # percentage of positive votes
+    positive_votes_percentage = (positive_proposal_votes / proposal.votes.count()) * 100
+    print('Positive votes percentage:', positive_votes_percentage)
+
+    if total_voted_community_members_percentage >= poll.quorum and positive_votes_percentage >= poll.approval_minimum:
+        print('Poll proposal has met approval and quorum')
+        poll.status = 1
```
Comment (Gofven) [inline flowback/poll/services/poll.py]: Setting poll.status = 1 means the poll has finished (which should only be set after poll finished counting)

File: flowback/poll/services/poll.py
Comment (Gofven) [inline flowback/poll/services/poll.py:1]: Remove prints if you don't use them anymore, these will litter the console, alternatively you can send debug log info to django using logging module https://docs.djangoproject.com/en/4.2/howto/logging/

File: flowback/poll/services/proposal.py [hunk 3]
```diff
@@ -90,3 +91,5 @@ def poll_proposal_priority_update(user_id: int, proposal_id: int, score: int) ->
 
     else:
         PollProposalPriority.objects.get(group_user=group_user, proposal=proposal).delete()
+
+    check_poll_proposal_met_approval_and_quorum(proposal_id=proposal_id)
```
Comment (Gofven) [inline flowback/poll/services/proposal.py]: This shouldn't be here (if it's checking poll_proposal), this should be counted at proposal_end_date, and iterate trough every proposal

File: flowback/poll/models.py [hunk 4]
```diff
@@ -46,8 +46,12 @@ class PollType(models.IntegerChoices):
     attachments = models.ForeignKey(FileCollection, on_delete=models.SET_NULL, null=True, blank=True)
     poll_type = models.IntegerField(choices=PollType.choices)
     parent = models.ForeignKey('self', on_delete=models.CASCADE, null=True, blank=True, related_name='children')
+
     quorum = models.IntegerField(default=None, null=True, blank=True,
                                  validators=[MinValueValidator(0), MaxValueValidator(100)])
+    approval_minimum = models.PositiveIntegerField(default=None, null=True, blank=True)
```
Comment (Gofven) [inline flowback/poll/models.py]: No checks found for approval minimum, must be below 100

File: flowback/poll/services/poll.py [hunk 5]
```diff
@@ -301,3 +308,50 @@ def poll_priority_update(user_id: int, poll_id: int, score: int) -> None:
...
+    positive_proposal_votes = proposal.positive_votes
+    print('Positive proposal votes:', positive_proposal_votes)
+
+    if poll.quorum is None:
+        print('No quorum set for poll')
+        return False
+    
+    if poll.approval_minimum is None:
+        print('No approval minimum set for poll')
+        return False
```
Comment (Gofven) [inline flowback/poll/services/poll.py]: Approval minimum will never pass unless set, either make it required or True

File: flowback/poll/services/poll.py [hunk 6]
```diff
@@ -301,3 +308,50 @@ def poll_priority_update(user_id: int, poll_id: int, score: int) -> None:
...
+    total_community_members = Group.objects.get(id=poll_community.id).groupuser_set.count()
+    print("Poll community:", poll_community)
+    print('Total community members:', total_community_members)
+
+    positive_proposal_votes = proposal.positive_votes
+    print('Positive proposal votes:', positive_proposal_votes)
+
+    if poll.quorum is None:
+        print('No quorum set for poll')
+        return False
```
Comment (Gofven) [inline flowback/poll/services/poll.py]: Quorum will never pass unless set, either make it required or True

File: flowback/poll/services/poll.py [hunk 7]
```diff
@@ -301,3 +308,50 @@ def poll_priority_update(user_id: int, poll_id: int, score: int) -> None:
 
     else:
         PollPriority.objects.get(group_user=group_user, poll=poll).delete()
+
+
+
+def check_poll_proposal_met_approval_and_quorum(*,proposal_id:int) -> bool:
+    """
+    Determines whether a poll proposal has achieved the necessary approval and quorum to become inactive.
```
Comment (Gofven) [inline flowback/poll/services/poll.py]: Active, not inactive
Also it's poll, not poll proposal

File: flowback/poll/services/proposal.py (hunk 3 above)
Comment (Gofven) [reply flowback/poll/services/proposal.py]: Additionally statement does nothing aside from changing wrong variable (see above)

File: flowback/poll/services/poll.py [hunk 8]
```diff
@@ -301,3 +308,50 @@ def poll_priority_update(user_id: int, poll_id: int, score: int) -> None:
 
     else:
         PollPriority.objects.get(group_user=group_user, poll=poll).delete()
+
+
+
+def check_poll_proposal_met_approval_and_quorum(*,proposal_id:int) -> bool:
```
Comment (Gofven) [inline flowback/poll/services/poll.py]: poll, not poll_proposal

File: flowback/poll/services/poll.py
Comment (Gofven) [reply flowback/poll/services/poll.py:1]: Additionally you seem to have mixed up poll_proposal and poll, i'm not sure if you wish to check approval_minimum for poll or proposal?

File: notes
Comment (Gofven) [inline notes:1]: Unnessesary file, delete or exclude from vcs

File: .DS_Store
Comment (Gofven) [inline .DS_Store:1]: Unnessesary, add to .gitignore

**Review by Gofven — CHANGES_REQUESTED** [review]
There's plenty of things that needs to be revised before accepting this pull request, although there are some nice additions i can see that you've added c:

File: notes
Comment (pvfarooq) [reply notes:1]: @narendersaini32 Any comments here?

File: flowback/poll/services/poll.py
Comment (pvfarooq) [reply flowback/poll/services/poll.py:1]: Yes, I was checking the approval minimum for the poll. The function name was misleading (renamed it). However, the way it works is correct.

File: flowback/poll/services/poll.py (hunk 1 above)
Comment (jayapal) [reply flowback/poll/services/poll.py]: This has been fixed by adding a active filter in the query.

Comment (pvfarooq) [reply flowback/poll/services/poll.py]: This has been fixed by adding an active filter in the query.

File: flowback/poll/services/poll.py (hunk 2 above)
Comment (pvfarooq) [reply flowback/poll/services/poll.py]: I've introduced a new poll status, '2', to represent the 'finalization_period'. When a poll meets the quorum and gains approval, its status will be updated to '2'. However, if a user downvotes the proposal and this results in the poll no longer meeting the quorum or approval criteria, the status will revert to '0' (ongoing).

File: flowback/poll/models.py (hunk 4 above)
Comment (pvfarooq) [reply flowback/poll/models.py]: I've added the min and max value validator for the field.

File: notes
Comment (pvfarooq) [reply notes:1]: Removed

File: flowback/poll/services/poll.py (hunk 6 above)
Comment (pvfarooq) [reply flowback/poll/services/poll.py]: Made the fields mandatory with a default value of 51 for approval and quorum

File: flowback/poll/services/poll.py (hunk 7 above)
Comment (pvfarooq) [reply flowback/poll/services/poll.py]: Updated

File: flowback/poll/services/poll.py (hunk 8 above)
Comment (pvfarooq) [reply flowback/poll/services/poll.py]: Updated

File: flowback/poll/services/poll.py (hunk 5 above)
Comment (pvfarooq) [reply flowback/poll/services/poll.py]: Updated

File: .DS_Store
Comment (pvfarooq) [reply .DS_Store:1]: Removed

**Review by Gofven — APPROVED** [review]

File: flowback/poll/services/proposal.py [hunk 9]
```diff
@@ -90,3 +91,60 @@ def poll_proposal_priority_update(user_id: int, proposal_id: int, score: int) ->
 
     else:
         PollProposalPriority.objects.get(group_user=group_user, proposal=proposal).delete()
+
+    check_poll_met_approval_and_quorum(proposal_id=proposal_id)
```
Comment (Gofven) [inline flowback/poll/services/proposal.py:95]: this should be checked elsewhere to ensure it'll be run reliably c:

**Pull request merged**